Mathlib/RingTheory/DedekindDomain/Factorization.lean

```diff
@@ -165,15 +190,374 @@ theorem finprod_heightOneSpectrum_factorization (I : Ideal R) (hI : I ≠ 0) :
+    ring
+
+/-- `val_v(v^n) = n` for every `n ∈ ℤ`. -/
+theorem count_zpow_self (n : ℤ) :
+    count K v ((v.asIdeal : FractionalIdeal R⁰ K) ^ n) = n := by
+  rw [count_zpow, count_self, mul_one]
+
+/-- If `v ≠ w` are two maximal ideals of `R`, then `val_v(w) = 0`. -/
+theorem count_maximal_coprime {w : HeightOneSpectrum R} (hw : w ≠ v) :
+    count K v (w.asIdeal : FractionalIdeal R⁰ K) = 0 := by
+  have hw_fact : (w.asIdeal : FractionalIdeal R⁰ K) =
+      spanSingleton R⁰ ((algebraMap R K) 1)⁻¹ * ↑w.asIdeal := by
+    rw [(algebraMap R K).map_one, inv_one, spanSingleton_one, one_mul]
+  have hw_ne_zero : (w.asIdeal : FractionalIdeal R⁰ K) ≠ 0 :=
+    coeIdeal_ne_zero.mpr w.ne_bot
+  have hv : Irreducible (Associates.mk v.asIdeal) := by apply v.associates_irreducible
+  have hw' : Irreducible (Associates.mk w.asIdeal) := by apply w.associates_irreducible
+  rw [count_well_defined K v hw_ne_zero hw_fact, Ideal.span_singleton_one, ← Ideal.one_eq_top,
+    Associates.mk_one, Associates.factors_one, Associates.count_zero hv, ofNat_zero, sub_zero,
+    natCast_eq_zero, ← pow_one (Associates.mk w.asIdeal), Associates.factors_prime_pow hw',
+    Associates.count_some hv, Multiset.replicate_one, Multiset.count_eq_zero,
+    Multiset.mem_singleton]
+  simp only [Subtype.mk.injEq]
+  rw [Associates.mk_eq_mk_iff_associated, associated_iff_eq, ← HeightOneSpectrum.ext_iff]
+  exact Ne.symm hw
+
+theorem count_maximal (w : HeightOneSpectrum R) :
+    count K v (w.asIdeal : FractionalIdeal R⁰ K) = if w = v then 1 else 0 := by
+  split_ifs with h
+  · rw [h, count_self]
+  · exact count_maximal_coprime K v h
+
+/-- `val_v(∏_{w ≠ v} w^{exps w}) = 0`. -/
+theorem count_finprod_coprime (exps : HeightOneSpectrum R → ℤ) :
+    count K v (∏ᶠ (w : HeightOneSpectrum R) (_ : w ≠ v),
+      (w.asIdeal : (FractionalIdeal R⁰ K)) ^ exps w) = 0 := by
+  apply finprod_mem_induction fun I => count K v I = 0
+  · exact count_one K v
+  · intro I I' hI hI'
+    by_cases h : I ≠ 0 ∧ I' ≠ 0
+    · rw [count_mul' K v, if_pos h, hI, hI', add_zero]
+    · rw [count_mul' K v, if_neg h]
+  · intro w hw
+    rw [count_zpow, count_maximal_coprime K v hw, MulZeroClass.mul_zero]
+
+/-- If `exps` is finitely supported, then `val_v(∏_w w^{exps w}) = exps v`. -/
+theorem count_finprod (exps : HeightOneSpectrum R → ℤ)
+    (h_exps : ∀ᶠ v : HeightOneSpectrum R in Filter.cofinite, exps v = 0) :
+    count K v (∏ᶠ v : HeightOneSpectrum R,
+      (v.asIdeal : FractionalIdeal R⁰ K) ^ exps v) = exps v := by
+  rw [finprod_eq_finset_prod_of_mulSupport_subset (s := h_exps.toFinset), count_prod]
+  · simp only [count_zpow, count_maximal, mul_ite, mul_one, mul_zero, Finset.sum_ite_eq',
+    Finite.mem_toFinset, mem_compl_iff, mem_setOf_eq, eq_comm (a := exps v), ite_not,
+    ite_eq_right_iff, imp_self]
+  · exact fun v hv ↦ zpow_ne_zero _ (coeIdeal_ne_zero.mpr v.ne_bot)
+  · rw [Finite.coe_toFinset]
+    intro v hv h
+    rw [mem_mulSupport, h, zpow_zero] at hv
+    exact hv (Eq.refl 1)
```

Comment:
```suggestion

theorem count_finsupp_prod (exps : HeightOneSpectrum R →₀ ℤ) :
    count K v (exps.prod (HeightOneSpectrum.asIdeal · ^ ·)) = exps v := by
  rw [Finsupp.prod, count_prod]
  · simp only [count_zpow, count_maximal, mul_ite, mul_one, mul_zero, Finset.sum_ite_eq',
      exps.mem_support_iff, ne_eq, ite_not, ite_eq_right_iff, @eq_comm ℤ 0, imp_self]
  · exact fun v hv ↦ zpow_ne_zero _ (coeIdeal_ne_zero.mpr v.ne_bot)

/-- If `exps` is finitely supported, then `val_v(∏_w w^{exps w}) = exps v`. -/
theorem count_finprod (exps : HeightOneSpectrum R → ℤ)
    (h_exps : ∀ᶠ v in Filter.cofinite, exps v = 0) :
    count K v (∏ᶠ v : HeightOneSpectrum R,
      (v.asIdeal : FractionalIdeal R⁰ K) ^ exps v) = exps v := by
  convert count_finsupp_prod K v (Finsupp.mk h_exps.toFinset exps (fun _ ↦ h_exps.mem_toFinset))
  rw [finprod_eq_finset_prod_of_mulSupport_subset (s := h_exps.toFinset), Finsupp.prod]
  · rfl
  · rw [Finite.coe_toFinset]
    intro v hv h
    rw [mem_mulSupport, h, zpow_zero] at hv
    exact hv (Eq.refl 1)

```
I was able to split this proof into two, which I think will be useful.

Reply:
Thanks!